### ISSUE TITLE
Add font-display to CSS file

### ISSUE THREAD

**Issue by vestimir** [rsms/inter#97]
Google Page Speed complains for fonts that block during loading. 

Details: https://developers.google.com/web/updates/2016/02/font-display

This can be fixed by adding the following code in https://github.com/rsms/inter/blob/master/docs/inter-ui.css

```css
font-display: auto;
```

Opening a PR for this :)

**Comment by rsms** [comment]
Doesn't this mean that some web browsers would show a fallback font, some would render the page with inivisble text (and some would wait for the correct font as is the case today)?

I.e. doing this might cause web sites out there using Inter UI to suddenly be displayed in other fonts for some duration of time, which would be unacceptable for aesthetically-oriented content where typography is a graphical cornerstone.

**Comment by vestimir** [comment]
Yep, that's exactly what will happen with this change on – instead of blocking, and not displaying text at all, the browser will display the system fallback until it loads the font files. The duration of this really depends on the particular website and it's setup. 

I guess it's a compromise of aestetics and performance score and people shouldn't be forced to either one by default. Another option will be to provide second CSS file for the non-blocking version so Page Speed wouldn't complain to people aiming at hitting 100/100. 

What do you think about that?

**Comment by vestimir** [comment]
@rsms and the more I think, the more I come to terms that putting options for a thing that works perfectly because it's simple is wrong. 

What about adding a simple instructions to the readme, or the FAQ on the website on the Page Speed subject instead?

**Comment by gabrielizaias** [comment]
@vestimir shouldn't it be `font-display: swap;` instead?

See: https://www.zachleat.com/web/comprehensive-webfonts/#font-display

**Comment by DominikSerafin** [comment]
@gabrielizaias, @vestimir, yep this should be probably `swap` if the goal is to always show some font before custom font loads (flash of unstyled text). The `auto` just lets the browser to decide the font-display strategy to use, and most of the time `block` strategy is used instead (flash of invisible text). 

I agree with @rsms that this probably should not be added to CDN stylesheet and maybe just referenced in docs/readme. Or at the very least, `text-display: auto` should be added, not `swap`. 

---

Hey BTW @rsms this is my first comment in this repo and I just wanted to thank you for this project. It's one of the most amazing project I've encountered on GitHub. The quality is just 💯.

**Comment by rsms** [comment]
Thank you @DominikSerafin :–)

**Comment by Addvilz** [comment]
We could have different CSS files instead. `swap` as default would be preferable to `auto` because of potential inconsistencies between browser implementations.

For display sensitive sites there could be CSS fallback file without `text-display` property as such.

Other than this, @rsms excellent job! I really love the font. Thank you very much!

**Comment by SoftCreatR** [comment]
@DominikSerafin You mean `font-display: auto`, not `text-display: auto` :)

Btw: Another good approach to prevent FOIT are local references. This requires the font (variants) to be installed on the local system, but if they are installed locally, no further roundtrips are required. Doing so doesn't hurt:

 `src: local("Inter Black"), local("Inter-Black"), ...` (using both naming cases, better safe than sorry)

However, there's a 3rd variant: Referencing the font group instead of a specific font variant:

`local("Inter")`

This one is used by Chrome for example. The problem with this variant is, that it doesn't work properly. It seems, that Chrome ignores the `font-weight` property and displays something, but not what you might expect. Dunno, if this is an intended behavior. Maybe, i'll fill a bug report for this.

TL;DR: Using locally installed versions of a font can greatly increase the site performance (Pagespeed doesn't care, but i don't care if Pagespeed cares) and there are no disadvantages, from what i can tell (except for the 3rd variant as explained above).

Regarding the `font-display` property: I personally use `swap` really often. Even in combination with Inter. I wouldn't use auto, because every browser does it's own thing then. This could lead in some unforseen behavior. So either don't use `font-display` at all, or use `swap`. Just my 2c.

**Comment by DominikSerafin** [comment]
@SoftCreatR you're right, thanks for letting me know. I've edited my comment.

**Comment by Calinou** [comment]
Should this issue be closed as #98 was closed?

**Comment by marvinrabe** [comment]
@Calinou No, because font-display is still not used in https://rsms.me/inter/inter.css

**Comment by acidoxee** [comment]
Any news on this matter? How are we supposed to set `font-display` when using the hosted version with `@import url('https://rsms.me/inter/inter.css');` ?

Seems not optimal to have to copy/paste the CSS locally and add a `font-display` property everywhere, especially since this prevents auto-updates of the font version and of the CSS file globally.

**Comment by DominikSerafin** [comment]
@acidoxee for now it seems hosting the font files locally is your only solution if you want to use `font-display`.

And adding some kind of default `font-display` to the `rsms.me` hosted files is the main point of discussion and issue here. Choosing a default value is not that straightforward and requires some compromises (see the first comments in this thread). 

May I ask what type of default `font-display` would you personally prefer here?

**Comment by acidoxee** [comment]
Since I would like to set it to `swap`, I guess I'll have to host the fonts locally and change the CSS.

It would be great though to have several values available through pre-made CSS files, once again for maintainability purposes on the consumers' side.

**Comment by marvinrabe** [comment]
Google Fonts supports font-display via URL parameter:
https://scotch.io/bar-talk/google-fonts-now-supports-font-display

@DominikSerafin Maybe this is the best solution for this issue aswell? Then there is no need to choose an optimal default. Keep "block" as default so it wont break compatibility. This method is already in use for choosing a different Inter version.

**Comment by originecode** [comment]
> Since I would like to set it to `swap`, I guess I'll have to host the fonts locally and change the CSS.
> 
> It would be great though to have several values available through pre-made CSS files, once again for maintainability purposes on the consumers' side.

Hey,

I think you missed something. You can do one thing that solve your problem. To use font-display, you just have to load a font, wherever it comes from. So, the trick is :
- create a local CSS file containing every rules of inter.css
- remove all version parameters (eg. : ?v=3.7)
- add your wished font-display property on each font face
- link this file to your project and your problem is solved.

**Comment by acidoxee** [comment]
> > Since I would like to set it to `swap`, I guess I'll have to host the fonts locally and change the CSS.
> > It would be great though to have several values available through pre-made CSS files, once again for maintainability purposes on the consumers' side.
> 
> Hey,
> 
> I think you missed something. You can do one thing that solve your problem. To use font-display, you just have to load a font, wherever it comes from. So, the trick is :
> 
> * create a local CSS file containing every rules of inter.css and add to font url the origin path to inter.css
> * remove all version parameters (eg. : ?v=3.7)
> * add your wished font-display property on each font face
> * link this file to your project and your problem is solved.

Hi Guillaume, Guillaume here 😉 

Sorry if my answer was not clear. What you are describing is exactly what I'm doing to set the `font-display` property.

My point was that it's not optimal for maintainability on our side (and every other customer's who has to do the same thing). Simply using something like `@import url('https://rsms.me/inter/inter.display-swap.css');` would enable our users to always benefit from the latest version of Inter, while taking away the burden of maintainability/manual updates from us.

As @marvinrabe mentioned, Google Fonts recently added the support for customizable `font-display` _via_ a simple `&display=swap` query string parameter. This is absolutely amazing, because from now on, we never ever need to maintain a CSS file while worrying that the font files we provide to our users are the latest available. All the more, if some "features" are added to the original CSS file that was copied, or if modifications are made to it, we wouldn't be immediately aware of them.

All this to say, this minor inconvenience is obviously not going to keep us from using the awesome Inter font, but helpers such as the ones discussed in this issue would be really great to have to replace any need for manual checks and updates.

**Comment by gabrielizaias** [comment]
I believe @rsms has final say in this matter.

The options are:
1. **Do nothing**: Consumers should re-host files and add `font-display` as their wish. Close PRs #176 and #177.
2. **Use `auto`**: Defer to browsers on how to handle it. Merge PR #176 and close PR #177
3. **Use `swap`**: Merge PR #177 and close PR #176 
4. **Add separate file for `swap`**? Proposed in a [comment](https://github.com/rsms/inter/issues/97#issuecomment-514408540). Hasn't been discussed at length yet.

**Comment by marvinrabe** [comment]
Another thing might be to publish the Font on Google Fonts. See here for details:

https://github.com/google/fonts/blob/master/CONTRIBUTING.md

**Comment by gabrielizaias** [comment]
@marvinrabe seems like it's nearly there? https://github.com/google/fonts/pull/1908

**Comment by rsms** [comment]
I think `swap` is the right move here for the hosted css file.
I've requested some small changes to #177 to make it mergeable.

**Comment by rsms** [comment]
Closed by #177
